### PR TITLE
Improve ignored check performance

### LINKED ISSUE #3967: Improve ignored check performance
### Description

First create a benchmark to compare possible solutions.

Our options allow ignoring certain things like transactions by name, spans by origin or checkins by slug. We are compiling the pattern each time which we could do only once at the start / when setting the option.

Ideas for improving:
- Iterate and only do `equals` comparison for all in list first, then iterate again and try to match via regex
- Compile pattern when setting the option, storing something like `FilterString` in options that holds both the String and the compiled pattern
- A decision cache for span origins only (as those should be low cardinality). We should not do this for transaction names or checkin slugs. Maybe a simple map where we store a bool for each origin we check can improve performance.

### PR DESCRIPTION
**Pull request by lbloder**
## :scroll: Description
Improves performance of checking whether certain SpanOrigins, TransactionNames or Checkins should be ignored.

## :bulb: Motivation and Context
Fixes #3967 
- Do equals check for all ignored items before regex check
- Use precompiled Pattern for regex check
- For SpanOrigins also add a decision cache 

## :pencil: Checklist

- [ ] I added tests to verify the changes.
- [ ] No new PII added or SDK only sends newly added PII if `sendDefaultPII` is enabled.
- [ ] I updated the docs if needed.
- [ ] I updated the wizard if needed.
- [ ] Review from the native team if needed.
- [ ] No breaking change or entry added to the changelog.
- [ ] No breaking change for hybrid SDKs or communicated to hybrid SDKs.

### CODE REVIEW

File: sentry/src/main/java/io/sentry/FilterString.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  @Override
+  public boolean equals(Object o) {
+    if (o == null || getClass() != o.getClass()) return false;
+    FilterString that = (FilterString) o;
+    return Objects.equals(filterString, that.filterString);
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hashCode(filterString);
```
Comment (adinauer) [inline sentry/src/main/java/io/sentry/FilterString.java:33]:
```suggestion
    return Objects.hash(filterString);
```

We seem to be using our own function here, not sure why but we should keep it consistent.

File: sentry/src/test/java/io/sentry/util/SpanUtilsTest.kt [hunk 2]
```diff
@@ -0,0 +1,49 @@
+package io.sentry.util
+
+import io.sentry.FilterString
+import kotlin.test.Test
+import kotlin.test.assertFalse
+import kotlin.test.assertTrue
+
+class SpanUtilsTest {
```
Comment (adinauer) [inline sentry/src/test/java/io/sentry/util/SpanUtilsTest.kt:8]: Should we test asking for the same origin again, so we can also test the cache?

**Review by adinauer — COMMENTED** [review]
Mostly LGTM already. Can you please add your performance comparison results?

**Review by adinauer — APPROVED** [review]

**Comment by lbloder** [comment]
# jmh benchmarks run on Java 17 (17.0.4.1 Temurin):
Benchmarks comparing the changes with our current implementation:
All Tests are run with 5 clean starts with 5 warmup runs and 5 actual benchmarking runs resulting in 25 benchmark runs per test. Each benchmark is run for 10 seconds and records the throughput achieved during its runtime.

Each implementation is given a list of 10 identical origins to be ignored.
For the original implementation this is a list of 10 strings, for the new implementation a list of 10 Strings with precompiled regex patterns.

## Comparison when no origin matches
All 10 items in the list need to be checked.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">


Benchmark | Mode | Cnt | Score |   | Error | Units | Result
-- | -- | -- | -- | -- | -- | -- | --
originalFilterNoMatches | thrpt | 25 | 790229,812 | ± | 5339,308 | ops/s | 100%
newFilterNoMatches | thrpt | 25 | 2992883037,803 | ± | 81483766,098 | ops/s | 378736%



</body>

</html>


## Comparison String equals
The 6th entry in the list matches using `String.equals`.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">


</head>

<body link="#467886" vlink="#96607D">


Benchmark | Mode | Cnt | Score |   | Error | Units | Result
-- | -- | -- | -- | -- | -- | -- | --
originalFilterEqualMatches | thrpt | 25 | 1573549,139 | ± | 38803,010 | ops/s | 100%
newFilterEqualsMatches | thrpt | 25 | 3024577381,724 | ± | 41660875,145 | ops/s | 192214%


</body>

</html>

## Comparison regex match
The 7th entry in the list matches using regex pattern.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip.htm">
<link rel=File-List
href="file:////Users/lukasbloder/Library/Group%20Containers/UBF8T346G9.Office/TemporaryItems/msohtmlclip/clip_filelist.xml">

</head>

<body link="#467886" vlink="#96607D">


Benchmark | Mode | Cnt | Score |   | Error | Units | Result
-- | -- | -- | -- | -- | -- | -- | --
originalFilterPatternMatches | thrpt | 25 | 1020669,472 | ± | 5584,216 | ops/s | 100%
newFilterPatternMatches | thrpt | 25 | 3024226052,507 | ± | 123600352,187 | ops/s | 296298%



</body>

</html>

**Pull request merged**